packages/core/src/browser/shell/tab-bar-decorator.ts

```diff
@@ -53,8 +55,16 @@ export class TabBarDecoratorService implements FrontendApplicationContribution {
     @inject(ContributionProvider) @named(TabBarDecorator)
     protected readonly contributions: ContributionProvider<TabBarDecorator>;
 
+    protected shell: ApplicationShell;
+    @inject(delayedShellInjectionFactory) protected readonly shellInjector: delayedShellInjectionFactory;
+
     initialize(): void {
         this.contributions.getContributions().map(decorator => decorator.onDidChangeDecorations(this.fireDidChangeDecorations));
+        // Avoid circular dependency caused by the direct injection of the ApplicationShell
+        setTimeout(() => {
+            this.shell = this.shellInjector();
+            this.shell.onDidChangeTrackableWidgets(() => this.fireDidChangeDecorations());
+        });
```

Comment:
This does not look good unfortunately.

Reply:
I didn't try but from the looks of it it should work without the `setTimeout`?

Reply:
It's worth a try, but I believe that `inversify` runs `@postConstruct()`-decorated methods before it returns the instance from the 'constructor', so that it will still fail without the `setTimeout` explicitly rescheduling the 'injection' of the `ApplicationShell` until after the `TabBarDecoratorService` is returned from its `inversify` instantiation routine.

Reply:
@colin-grant-work I reverted the commit, Is the PR ready to merge now?